applications/IgaApplication/custom_elements/shell_3p_element.cpp

```diff
@@ -0,0 +1,766 @@
+        array_1d<double, 3> a_contravariant_2 = rKinematicVariables.a1*a_ab_contravariant[2] + rKinematicVariables.a2*a_ab_contravariant[1];
+
+
+        //Local cartesian coordinates
+        double l_a1 = norm_2(rKinematicVariables.a1);
+        array_1d<double, 3> e1 = rKinematicVariables.a1 / l_a1;
+        double l_a_contravariant_2 = norm_2(a_contravariant_2);
+        array_1d<double, 3> e2 = a_contravariant_2 / l_a_contravariant_2;
+
+        // e * a_contravariant
+        Matrix G = ZeroMatrix(2, 2);
+        G(0, 0) = inner_prod(e1, a_contravariant_1);
+        G(0, 1) = inner_prod(e1, a_contravariant_2);
+        G(1, 0) = inner_prod(e2, a_contravariant_1);
+        G(1, 1) = inner_prod(e2, a_contravariant_2);
+
+        //Transformation matrix T
+        if (rT.size1() != 3 && rT.size2() != 3)
+            rT.resize(3, 3);
+
+        rT = ZeroMatrix(3, 3);
+        rT(0, 0) = pow(G(0, 0), 2);
+        rT(0, 1) = pow(G(0, 1), 2);
+        rT(0, 2) = 2 * G(0, 0) * G(0, 1);
+
+        rT(1, 0) = pow(G(1, 0), 2);
+        rT(1, 1) = pow(G(1, 1), 2);
+        rT(1, 2) = 2 * G(1, 0) * G(1, 1);
+
+        rT(2, 0) = 2 * G(0, 0) * G(1, 0);
+        rT(2, 1) = 2 * G(0, 1) * G(1, 1);
+        rT(2, 2) = 2 * (G(0, 0) * G(1, 1) + G(0, 1) * G(1, 0));
+    }
+
+    void Shell3pElement::CalculateConstitutiveVariables(
+        IndexType IntegrationPointIndex,
+        KinematicVariables& rActualKinematic,
+        ConstitutiveVariables& rThisConstitutiveVariablesMembrane,
+        ConstitutiveVariables& rThisConstitutiveVariablesCurvature,
+        ConstitutiveLaw::Parameters& rValues,
+        const ConstitutiveLaw::StressMeasure ThisStressMeasure
+    )
+    {
+        rValues.GetOptions().Set(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN, true);
+        rValues.GetOptions().Set(ConstitutiveLaw::COMPUTE_STRESS);
+        rValues.GetOptions().Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR);
+
+        array_1d<double, 3> strain_vector = 0.5 * (rActualKinematic.a_ab_covariant - m_A_ab_covariant_vector[IntegrationPointIndex]);
+        rThisConstitutiveVariablesMembrane.StrainVector = prod(m_T_vector[IntegrationPointIndex], strain_vector);
+
+        array_1d<double, 3> curvature_vector = rActualKinematic.b_ab_covariant - m_B_ab_covariant_vector[IntegrationPointIndex];
+        rThisConstitutiveVariablesCurvature.StrainVector = prod(m_T_vector[IntegrationPointIndex], curvature_vector);
+
+        // Constitive Matrices DMembrane and DCurvature
+        rValues.SetStrainVector(rThisConstitutiveVariablesMembrane.StrainVector); //this is the input parameter
+        rValues.SetStressVector(rThisConstitutiveVariablesMembrane.StressVector);    //this is an ouput parameter
+        rValues.SetConstitutiveMatrix(rThisConstitutiveVariablesMembrane.ConstitutiveMatrix); //this is an ouput parameter
+
+        m_constitutive_law_vector[0]->CalculateMaterialResponse(rValues, ThisStressMeasure);
```

Comment:
why only for one?

Reply:
That's a bug. That might have been tricky to found out...

Thanks!